src/main/java/org/jetbrains/bsp/bazel/server/DiagnosticsDispatcher.java

```diff
@@ -0,0 +1,134 @@
+package org.jetbrains.bsp.bazel.server;
+
+import ch.epfl.scala.bsp4j.BuildClient;
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
+import ch.epfl.scala.bsp4j.Diagnostic;
+import ch.epfl.scala.bsp4j.DiagnosticSeverity;
+import ch.epfl.scala.bsp4j.Position;
+import ch.epfl.scala.bsp4j.PublishDiagnosticsParams;
+import ch.epfl.scala.bsp4j.Range;
+import ch.epfl.scala.bsp4j.SourceItem;
+import ch.epfl.scala.bsp4j.TextDocumentIdentifier;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import io.bazel.rules_scala.diagnostics.Diagnostics;
+import io.bazel.rules_scala.diagnostics.Diagnostics.Severity;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.bsp.bazel.common.Uri;
+
+public class DiagnosticsDispatcher {
+
+  private static final Logger LOGGER = LogManager.getLogger(DiagnosticsDispatcher.class);
+
+  private final Map<Severity, DiagnosticSeverity> CONVERTED_SEVERITY =
+      new ImmutableMap.Builder<Diagnostics.Severity, DiagnosticSeverity>()
+          .put(Severity.UNKNOWN, DiagnosticSeverity.ERROR)
+          .put(Severity.ERROR, DiagnosticSeverity.ERROR)
+          .put(Severity.WARNING, DiagnosticSeverity.WARNING)
+          .put(Severity.INFORMATION, DiagnosticSeverity.INFORMATION)
+          .put(Severity.HINT, DiagnosticSeverity.HINT)
+          .build();
+
+  private final BazelBspServer bspServer;
+  private final BuildClient bspClient;
+
+  public DiagnosticsDispatcher(BazelBspServer bspServer, BuildClient bspClient) {
+    this.bspServer = bspServer;
+    this.bspClient = bspClient;
+  }
+
+  public void collectDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics,
+      BuildTargetIdentifier target,
+      String diagnosticsLocation)
+      throws IOException {
+    Diagnostics.TargetDiagnostics targetDiagnostics =
+        Diagnostics.TargetDiagnostics.parseFrom(Files.readAllBytes(Paths.get(diagnosticsLocation)));
+
+    for (Diagnostics.FileDiagnostics fileDiagnostics : targetDiagnostics.getDiagnosticsList()) {
+      LOGGER.info("Inserting diagnostics for path: {}", fileDiagnostics.getPath());
+
+      filesToDiagnostics.put(
+          getUriForPath(fileDiagnostics.getPath()), convertDiagnostics(target, fileDiagnostics));
+    }
+  }
+
+  public void emitDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics, BuildTargetIdentifier target) {
+    for (SourceItem source : bspServer.getCachedBuildTargetSources(target)) {
+      Uri sourceUri = Uri.fromFileUri(source.getUri());
+      if (!filesToDiagnostics.containsKey(sourceUri)) {
+        addSource(filesToDiagnostics, sourceUri, target);
+      }
+
+      if (bspClient != null) {
```

Comment:
can bspClient ever be null? does it make sense to check it at this point?

Reply:
Yeah, I also thought it didn't, I'll remove it.